bot/admin/server/src/main/kotlin/BotAdminService.kt

```diff
@@ -1141,23 +1145,30 @@ object BotAdminService {
         }
 
         // delete the RAG configuration
-        ragConfigurationDAO.findByNamespaceAndBotId(app.namespace, app.name)?.let {
-            ragConfigurationDAO.delete(it._id)
+        ragConfigurationDAO.findByNamespaceAndBotId(app.namespace, app.name)?.let { config ->
+            ragConfigurationDAO.delete(config._id)
+            config.llmSetting.apiKey?.let { SecurityUtils.deleteSecret(it) }
+            config.emSetting.apiKey?.let { SecurityUtils.deleteSecret(it) }
         }
 
         // delete the Sentence Generation configuration
-        sentenceGenerationConfigurationDAO.findByNamespaceAndBotId(app.namespace, app.name)?.let {
-            sentenceGenerationConfigurationDAO.delete(it._id)
+        sentenceGenerationConfigurationDAO.findByNamespaceAndBotId(app.namespace, app.name)?.let { config ->
+            sentenceGenerationConfigurationDAO.delete(config._id)
+            config.llmSetting.apiKey?.let { SecurityUtils.deleteSecret(it) }
         }
 
         // delete the Observability configuration
-        observabilityConfigurationDAO.findByNamespaceAndBotId(app.namespace, app.name)?.let {
-            observabilityConfigurationDAO.delete(it._id)
+        observabilityConfigurationDAO.findByNamespaceAndBotId(app.namespace, app.name)?.let { config ->
+            observabilityConfigurationDAO.delete(config._id)
+            config.setting.takeIf { it is LangfuseObservabilitySetting }?.let {
+                SecurityUtils.deleteSecret((config.setting as LangfuseObservabilitySetting).secretKey)
+            }
         }
 
         // delete the Vector Store configuration
-        vectorStoreConfigurationDAO.findByNamespaceAndBotId(app.namespace, app.name)?.let {
-            vectorStoreConfigurationDAO.delete(it._id)
+        vectorStoreConfigurationDAO.findByNamespaceAndBotId(app.namespace, app.name)?.let { config ->
+            vectorStoreConfigurationDAO.delete(config._id)
+            SecurityUtils.deleteSecret(config.setting.password)
```

Comment:
At which layer are the Mongo transaction handled ? If the secret deletion did not happen, will the mongo deletion be rollback ? What happens if we replay this deletion process because ot partially failed ? Is it idempotent ?

Reply:
I knew the operation could succeed on the database, but not on the cloud, but I didn't think it was a big deal, because you'd end up with an orphaned secret. What do you think @Benvii? Is it necessary to go back (redo an insertion) on the MongoDB deletion if the secret deletion didn't go well?

Reply:
Good point @AurelienSylvan I agree that it would be cleaner to make a rollback, the deletion fail is logged I think, if we have a good monitoring of warning / errors logs it enough for me.